[doc] Update performance.md (#6911)
taichi-dev/taichi: docs/lang/articles/performance_tuning/performance.md

## For-loop decorators

> As discussed in previous topics, Taichi kernels automatically parallelize for-loops in the outermost scope. Our compiler sets the settings automatically to best explore the target architecture. Nonetheless, for Ninjas seeking the final few percent of speed, we give several APIs to allow developers to fine-tune their programmes. Specifying a proper `block dim`, for example, might result in a nearly 3x speed gain in [examples/mpm3d.py](https://github.com/taichi-dev/taichi/blob/master/python/taichi/examples/mpm3d.py).

Reviewer: ```suggestion
As discussed in previous topics, Taichi kernels automatically parallelize for-loops in the outermost scope. Our compiler sets the settings automatically to best explore the target architecture. Nonetheless, for Ninjas seeking the final few percent of speed, we give several APIs to allow developers to fine-tune their programs. Specifying a proper `block dim`, for example, might result in a nearly 3x speed gain in [examples/mpm3d.py](https://github.com/taichi-dev/taichi/blob/master/python/taichi/examples/mpm3d.py).
```
Let's use American English here~

### Background: Thread hierarchy of GPUs

It is worthy to quickly discuss the **thread hierarchy** on contemporary GPU architectures in order to help you understand how the previously mentioned for-loop is parallelized.

> From fine-grained to coarse-grained, the computation units are as follows: **iteration** **thread** **block** **grid**.

Reviewer: ```suggestion
From fine-grained to coarse-grained, the computation units are as follows: **iteration**, **thread**, **block**, **grid**.
```

- **iteration**: The **body of a for-loop** is an iteration. Each iteration corresponds to a different I value in the for-loop.

> - **thread**: Iterations are classified as threads. A thread is the smallest parallelized unit. All iterations inside a thread are **serial** in nature. To maximise parallel efficiency, we normally employ one iteration per thread.
> - **block**: Threads are organised into groups called blocks. **Parallel** execution is used for all threads within a block. Threads within a block can share **block local storage**.

Reviewer: ```suggestion
- **thread**: Iterations are classified as threads. A thread is the smallest parallelized unit. All iterations inside a thread are **serial** in nature. To maximize parallel efficiency, we normally employ one iteration per thread.
- **block**: Threads are organized into groups called blocks. **Parallel** execution is used for all threads within a block. Threads within a block can share **block local storage**.
```

![](../static/assets/bls_indices_mapping.png)

You do not need to be concerned about these fundamental elements as a user.
Taichi automatically does all inference and global/block-local mapping.
That is, Taichi will preallocate a CUDA shared memory buffer of size `5x6`, preload `a`'s contents into this buffer, then replace all `a` (global memory) accesses with the buffer in the loop body. While this basic example does not change `a` itself, if a block-cached field is written, Taichi will produce code that returns the buffer to global memory.

:::note
BLS does not come cheap. Remember that BLS is intended for stencil computations with a high number of overlapping global memory accesses. If this is not the case, pre-loading and post-storing may actually degrade performance.

Furthermore, recent generations of Nvidia GPU cards have closed the read-only access gap between global memory and shared memory. Currently, we discovered that BLS is more effective for storing the destinations of atomic actions.

> As a general rule of thumb, we recommnend running benchmarks to determine whether or not you should enable BLS.

Reviewer: ```suggestion
As a general rule of thumb, we recommend running benchmarks to determine whether or not you should enable BLS.
```
